File: 백준/Silver/14940. 쉬운 최단거리/쉬운 최단거리.js
const filePath = process.platform === 'linux' ? '/dev/stdin' : 'e.txt';
const input = require('fs')
	.readFileSync(filePath)
	.toString()
	.trim()
	.split('\n');
const [row, col] = input.shift().split(' ').map(Number);

const [WALL, GO, GOAL] = [0, 1, 2];
const DX = [0, 1, 0, -1];
const DY = [1, 0, -1, 0];

const stepMap = new Array(row).fill().map(() => new Array(col).fill(Infinity));
const map = new Array(row).fill();

for (let i = 0; i < row; i++) {
	map[i] = input[i].split(' ').map(Number);
}

const bfs = (start) => {
	const queue = [];

	stepMap[start[0]][start[1]] = 0;
	queue.push([...start, 0]);

	while (queue.length) {
		const [y, x, steps] = queue.shift();

		for (let i = 0; i < DX.length; i++) {
			const [ny, nx] = [y + DY[i], x + DX[i]];

			// 모서리 아니고, 길 맞고, 이 경로가 더 짧게 도달하는 경우
			// if (ny === 14 && nx === 14) console.log(map[ny][nx] === GO);
			if (
				![-1, row].includes(ny) &&
				![-1, col].includes(nx) &&
				map[ny][nx] === GO &&
				stepMap[ny][nx] > steps + 1
			) {
				stepMap[ny][nx] = steps + 1;
				queue.push([ny, nx, steps + 1]);
			}
		}
	}
};
// 종착지는 0으로 해야한다.

for (let y = 0; y < row; y++) {
	for (let x = 0; x < col; x++) {
		if (map[y][x] === GOAL) {
			bfs([y, x]);
			break;
		}
	}
}

for (let y = 0; y < row; y++) {
	for (let x = 0; x < col; x++) {
		if (map[y][x] === WALL) stepMap[y][x] = 0;
		if (stepMap[y][x] === Infinity) stepMap[y][x] = -1;
	}
}

console.log(stepMap.map((v) => v.join(' ')).join('\n'));

File: 백준/Silver/14940. 쉬운 최단거리/쉬운 최단거리.py
'''
BFS, 다익 둘 다 가능하다. 
시간복잡도는 모르겠다.  
'''

'''
from collections import deque
q = deque(배열)
'''

START = 2
FLOOR = 1
WALL = 0

from collections import deque

col, row = map(int, input().split())
mp = []
for _ in range(col):
    line = list(map(int, input().split()))
    mp.append(line)
    
# 시작 지점 찾기
start = None
isBreak = False;
for y in range(col):
    for x in range(row):
        if mp[y][x] == START:
            start = (y, x)
            isBreak = True
            break
    if isBreak:
        break;
        
DX = [0, 1, 0, -1]
DY = [1, 0, -1, 0]

visited = [([False] * row) for _ in range(col)] # 숫자는 방문 False는 미방문
visited[start[0]][start[1]] = 0

def bfs():
    q = deque()
    q.append((start[0], start[1], 0)) # y, x, 거리
        
    while len(q) > 0:
        y, x, c = q.popleft()
            
        for i in range(len(DX)):
            ny, nx = y + DY[i], x + DX[i]
            
            # 범위 이탈
            if ny < 0 or ny >= col or nx < 0 or nx >= row:
                continue;
            
            # 진입 불가
            if mp[ny][nx] != FLOOR:
                continue

            # 방문 기록 존재하며 그 기록이 현재보다 빠를 경우
            if visited[ny][nx] != False and c + 1 >= visited[ny][nx]:
                continue
                
            newCost = c + 1
            q.append((ny, nx, newCost))
            visited[ny][nx] = newCost

bfs();


for y in range(col):
    for x in range(row):
        if mp[y][x] == WALL: 
            visited[y][x] = 0
            continue;
        if mp[y][x] == FLOOR and visited[y][x] == False:
            visited[y][x] = -1
            


for y in range(col):
    print(' '.join(map(str, visited[y])))
    
    
    
    
    
    
        
